{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "85993d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dae00e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_merged_summary=\"/home/lmf/MOSAIC/ORGANIC/05_vOTUs/checkV_merged_quality_summary.tot.txt\"\n",
    "# input_cluster_file=\"/home/lmf/MOSAIC/ORGANIC/05_vOTUs/combined_positive_viral_contigs.tot_95-85.clstr\"\n",
    "# output_representatives=\"/home/lmf/MOSAIC/ORGANIC/05_vOTUs/vOTU_clustering_rep_list.csv\"\n",
    "# output_checkv_categories=\"/home/lmf/MOSAIC/ORGANIC/05_vOTUs/vOTU_clustering_rep_list_checkv_per_category.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ff1e2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_merged_summary=snakemake.input.merged_summary\n",
    "input_cluster_file=snakemake.input.cluster_file\n",
    "output_representatives=snakemake.output.representatives\n",
    "output_checkv_categories=snakemake.output.checkv_categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f3218f32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id_vOTU</th>\n",
       "      <th>contig_length_vOTU</th>\n",
       "      <th>checkv_quality_vOTU</th>\n",
       "      <th>completeness_vOTU</th>\n",
       "      <th>provirus_vOTU</th>\n",
       "      <th>warnings_vOTU</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124158</th>\n",
       "      <td>Vir_3_T_tig00004100_len=1016_reads=2_class=con...</td>\n",
       "      <td>718.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97441</th>\n",
       "      <td>Vir_3_2_2_NODE_9072_length_1000_cov_2.717460</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Low-quality</td>\n",
       "      <td>2.16</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37074</th>\n",
       "      <td>Vir_3_T_NODE_45770_length_1000_cov_3.016931</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Low-quality</td>\n",
       "      <td>2.37</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37075</th>\n",
       "      <td>Vir_3_T_NODE_45771_length_1000_cov_3.003175</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37076</th>\n",
       "      <td>Vir_3_T_NODE_45772_length_1000_cov_2.996825</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Low-quality</td>\n",
       "      <td>0.89</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125779</th>\n",
       "      <td>Vir_3_T_tig00022782_len=123915_reads=429_class...</td>\n",
       "      <td>125742.0</td>\n",
       "      <td>Medium-quality</td>\n",
       "      <td>88.97</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vir_3_T_NODE_4_length_136511_cov_14.214501</td>\n",
       "      <td>136511.0</td>\n",
       "      <td>Low-quality</td>\n",
       "      <td>37.63</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125298</th>\n",
       "      <td>Vir_3_T_tig00000162_len=150443_reads=3504_clas...</td>\n",
       "      <td>151838.0</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>100.00</td>\n",
       "      <td>No</td>\n",
       "      <td>contig &gt;1.5x longer than expected genome length</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125083</th>\n",
       "      <td>Vir_3_T_tig00000093_len=165649_reads=3063_clas...</td>\n",
       "      <td>164373.0</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>100.00</td>\n",
       "      <td>No</td>\n",
       "      <td>contig &gt;1.5x longer than expected genome length</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123085</th>\n",
       "      <td>Vir_3_T_tig00000004_len=184625_reads=1555_clas...</td>\n",
       "      <td>186223.0</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>100.00</td>\n",
       "      <td>No</td>\n",
       "      <td>contig &gt;1.5x longer than expected genome length</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53550 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           contig_id_vOTU  contig_length_vOTU  \\\n",
       "124158  Vir_3_T_tig00004100_len=1016_reads=2_class=con...               718.0   \n",
       "97441        Vir_3_2_2_NODE_9072_length_1000_cov_2.717460              1000.0   \n",
       "37074         Vir_3_T_NODE_45770_length_1000_cov_3.016931              1000.0   \n",
       "37075         Vir_3_T_NODE_45771_length_1000_cov_3.003175              1000.0   \n",
       "37076         Vir_3_T_NODE_45772_length_1000_cov_2.996825              1000.0   \n",
       "...                                                   ...                 ...   \n",
       "125779  Vir_3_T_tig00022782_len=123915_reads=429_class...            125742.0   \n",
       "0              Vir_3_T_NODE_4_length_136511_cov_14.214501            136511.0   \n",
       "125298  Vir_3_T_tig00000162_len=150443_reads=3504_clas...            151838.0   \n",
       "125083  Vir_3_T_tig00000093_len=165649_reads=3063_clas...            164373.0   \n",
       "123085  Vir_3_T_tig00000004_len=184625_reads=1555_clas...            186223.0   \n",
       "\n",
       "       checkv_quality_vOTU  completeness_vOTU provirus_vOTU  \\\n",
       "124158      Not-determined                NaN            No   \n",
       "97441          Low-quality               2.16            No   \n",
       "37074          Low-quality               2.37            No   \n",
       "37075       Not-determined                NaN            No   \n",
       "37076          Low-quality               0.89            No   \n",
       "...                    ...                ...           ...   \n",
       "125779      Medium-quality              88.97            No   \n",
       "0              Low-quality              37.63            No   \n",
       "125298        High-quality             100.00            No   \n",
       "125083        High-quality             100.00            No   \n",
       "123085        High-quality             100.00            No   \n",
       "\n",
       "                                          warnings_vOTU  \n",
       "124158                          no viral genes detected  \n",
       "97441                                               NaN  \n",
       "37074                           no viral genes detected  \n",
       "37075                                               NaN  \n",
       "37076                                               NaN  \n",
       "...                                                 ...  \n",
       "125779                                              NaN  \n",
       "0                                                   NaN  \n",
       "125298  contig >1.5x longer than expected genome length  \n",
       "125083  contig >1.5x longer than expected genome length  \n",
       "123085  contig >1.5x longer than expected genome length  \n",
       "\n",
       "[53550 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read and preprocess the clustered_df DataFrame\n",
    "clustered_df = pd.read_csv(input_cluster_file, sep=\"\\t\", names=[\"rep\", \"mem\"])\n",
    "clustered_df[\"mem\"] = clustered_df[\"mem\"].str.split(\",\")\n",
    "clustered_df = clustered_df.explode('mem').reset_index(drop=True)\n",
    "\n",
    "# Read and preprocess the checkv_df_vOTUs DataFrame\n",
    "checkv_df_vOTUs = pd.read_csv(input_merged_summary, sep=\"\\t\",\n",
    "                            names= [\"contig_id_vOTU\", \"contig_length_vOTU\", \"provirus_vOTU\", \"proviral_length_vOTU\",\n",
    "                           \"gene_count_vOTU\", \"viral_genes_vOTU\", \"host_genes_vOTU\", \"checkv_quality_vOTU\",\n",
    "                           \"miuvig_quality_vOTU\", \"completeness_vOTU\", \"completeness_method_vOTU\",\n",
    "                           \"contamination_vOTU\", \"kmer_freq_vOTU\", \"warnings_vOTU\"])\n",
    "checkv_df_vOTUs = checkv_df_vOTUs[[\"contig_id_vOTU\", \"contig_length_vOTU\", \"checkv_quality_vOTU\",\n",
    "                                   \"completeness_vOTU\", \"provirus_vOTU\", \"warnings_vOTU\"]]\n",
    "checkv_df_vOTUs[\"contig_length_vOTU\"] = checkv_df_vOTUs[\"contig_length_vOTU\"].astype(float)\n",
    "checkv_df_vOTUs[\"completeness_vOTU\"] = checkv_df_vOTUs[\"completeness_vOTU\"].astype(float)\n",
    "\n",
    "# Create copies of the checkv_df_vOTUs DataFrame\n",
    "checkv_df_rep = checkv_df_vOTUs.copy()\n",
    "checkv_df_rep.columns = [\"contig_id_rep\", \"contig_length_rep\", \"checkv_quality_rep\", \"completeness_rep\",\n",
    "                         \"provirus_rep\", \"warnings_rep\"]\n",
    "\n",
    "# Merge DataFrames and drop unnecessary columns\n",
    "merged_df = clustered_df.merge(checkv_df_rep, left_on=\"rep\", right_on=\"contig_id_rep\", how=\"left\") \\\n",
    "                        .merge(checkv_df_vOTUs, left_on=\"mem\", right_on=\"contig_id_vOTU\", how=\"left\")\n",
    "merged_df = merged_df.drop(['contig_id_vOTU', 'contig_id_rep'], axis=1)\n",
    "merged_df[['completeness_vOTU', 'contig_length_vOTU', 'completeness_rep']] = merged_df[\n",
    "    ['completeness_vOTU', 'contig_length_vOTU', 'completeness_rep']].fillna(0)\n",
    "merged_df[\"completeness_vOTU\"] = merged_df[\"completeness_vOTU\"].astype(float)\n",
    "merged_df[\"contig_length_vOTU\"] = merged_df[\"contig_length_vOTU\"].astype(float)\n",
    "\n",
    "# Choose the contig with the highest completeness and the lowest contig length\n",
    "merged_df_best = merged_df[merged_df.groupby(['rep'], sort=False)['completeness_vOTU'].transform(max) == merged_df['completeness_vOTU']]\n",
    "merged_df_best = merged_df_best[merged_df_best.groupby(['rep'], sort=False)['contig_length_vOTU'].transform(min) == merged_df_best['contig_length_vOTU']]\n",
    "merged_df_best = merged_df_best.groupby(['rep']).first().reset_index()\n",
    "\n",
    "# Select final singletons and grouped contigs based on checkv quality\n",
    "final_singletons = merged_df_best[merged_df_best[\"rep\"] == merged_df_best[\"mem\"]][\"rep\"].tolist()\n",
    "\n",
    "merged_df_diff = merged_df_best[merged_df_best[\"rep\"] != merged_df_best[\"mem\"]]\n",
    "merged_df_diff = merged_df_diff.replace({'Complete': 4, 'High-quality': 3, 'Medium-quality': 2,'Low-quality': 1, 'Not-determined': 1})\n",
    "\n",
    "# Choose representatives based on checkv quality comparison\n",
    "merged_df_greatherthan = merged_df_diff[merged_df_diff[\"checkv_quality_rep\"] >= merged_df_diff[\"checkv_quality_vOTU\"]]\n",
    "gr1 = merged_df_greatherthan[\"rep\"].tolist()\n",
    "\n",
    "merged_df_smallerthan = merged_df_diff[merged_df_diff[\"checkv_quality_rep\"] < merged_df_diff[\"checkv_quality_vOTU\"]]\n",
    "gr2 = merged_df_smallerthan[\"mem\"].tolist()\n",
    "\n",
    "representatives = final_singletons + gr1 + gr2\n",
    "final_df_rep = checkv_df_vOTUs[checkv_df_vOTUs['contig_id_vOTU'].isin(representatives)]\n",
    "\n",
    "# Save representative contig IDs to a CSV file\n",
    "final_df_rep[\"contig_id_vOTU\"].to_csv(output_representatives, index=False, header=False)\n",
    "\n",
    "# Sort the final_df_rep DataFrame by contig_length_vOTU column\n",
    "final_df_rep.sort_values(by=\"contig_length_vOTU\")\n",
    "\n",
    "                                                                  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5846bd6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>checkv_quality_vOTU</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Complete</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>High-quality</th>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low-quality</th>\n",
       "      <td>46362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Medium-quality</th>\n",
       "      <td>360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not-determined</th>\n",
       "      <td>6593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         0\n",
       "checkv_quality_vOTU       \n",
       "Complete                54\n",
       "High-quality           181\n",
       "Low-quality          46362\n",
       "Medium-quality         360\n",
       "Not-determined        6593"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df_rep.groupby(\"checkv_quality_vOTU\").size().to_csv(output_checkv_categories, index=False, header=False)\n",
    "final_df_rep.groupby(\"checkv_quality_vOTU\").size().to_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee7ff45e",
   "metadata": {},
   "source": [
    "# BANREP"
   ]
  },
  {
   "cell_type": "raw",
   "id": "cc9e72d0",
   "metadata": {},
   "source": [
    "grep -c \">\" vOT*#COMBINE vOTUS WITH M AND H QUAL\n",
    "cat subassembly_vOTU_clustering_rep.fasta BanRep_contigs.fasta > vOTUs_and_BanRep.fasta\n",
    "\n",
    "#CLUSTER AT SPECIES LEVEL\n",
    "makeblastdb -in vOTUs_and_BanRep.fasta -dbtype nucl -out vOTUs_and_BanRep.fasta\n",
    "blastn -query vOTUs_and_BanRep.fasta -db vOTUs_and_BanRep.fasta -outfmt '6 std qlen slen' -max_target_seqs 10000 -out vOTUs_and_BanRep.blastout -num_threads 144 #14:52-16:36 = 1h,44min\n",
    "\n",
    "python /home/lmf/Mosaic/mosaic/scripts/anicalc_checkv.py  -i vOTUs_and_BanRep.blastout -o vOTUs_and_BanRep.aniout\n",
    "python /home/lmf/Mosaic/mosaic/scripts/aniclust_checkv.py --fna vOTUs_and_BanRep.fasta --ani vOTUs_and_BanRep.aniout --out vOTUs_and_BanRep.txt --min_ani 95 --min_tcov 85 --min_qcov 0    \n",
    "cut  -f1 vOTUs_and_BanRep.txt > vOTUs_and_BanRep_rep_list.txt\n",
    "seqtk subseq vOTUs_and_BanRep.fasta vOTUs_and_BanRep_rep_list.txt > vOTUs_and_BanRep_rep.fasta\n",
    "grep -c \">\" vOTUs_and_BanRep_rep.fasta #111051\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32ebb5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# merged_df\n",
    "merged_df=merged_df[merged_df[\"rep\"]!=merged_df[\"mem\"]]\n",
    "\n",
    "merged_df=merged_df[merged_df[\"completeness_rep\"]!=merged_df[\"completeness_vOTU\"]]\n",
    "merged_df=merged_df[merged_df[\"completeness_rep\"]>merged_df[\"completeness_vOTU\"]+10]\n",
    "\n",
    "merged_df\n",
    "# clustered_df=clustered_df.merge(checkv_df_rep, left_on=\"rep\", right_on=\"contig_id_rep\", how=\"left\")\n",
    "# clustered_df = clustered_df.drop('contig_id_rep', axis=1)\n",
    "# clustered_df=clustered_df[clustered_df[\"contig_length_rep\"].isna()==False]\n",
    "\n",
    "\n",
    "# merged_df\n",
    "\n",
    "# # merged_df.groupby('rep').first().groupby('checkv_quality_rep').count()\n",
    "# merged_df=merged_df[(merged_df[\"completeness_vOTU\"]<merged_df[\"completeness_rep\"])]\n",
    "# merged_df\n",
    "# merged_df.groupby('checkv_quality_rep').count()\n",
    "\n",
    "# merged_df=merged_df.replace('Complete',4)\n",
    "# merged_df=merged_df.replace('High-quality',3)\n",
    "# merged_df=merged_df.replace('Medium-quality',2)\n",
    "# merged_df=merged_df.replace('Low-quality',1)\n",
    "# merged_df=merged_df.replace('Not-determined',0)\n",
    "# merged_df['checkv_quality_votu'] =  merged_df['checkv_quality_votu'].fillna(4)\n",
    "# merged_df\n",
    "# # merged_df['warnings'] =  merged_df['warnings'].fillna(\"none\")\n",
    "\n",
    "# # # merged_df\n",
    "# # merged_df[merged_df[\"warnings\"]!=\"none\"]\n",
    "# # merged_df=merged_df[merged_df[\"warnings\"]==\"none\"]\n",
    "\n",
    "\n",
    "textfile = open(\"/home/lmf/PhylloVir/VIRAL_WORLD/VIRAL_FRACTION/08_ASSEMBLY_TEST/reference_genomes.txt\", \"w\")\n",
    "merged_df\n",
    "for element in list(merged_df[\"rep\"]):\n",
    "    textfile.write(element + \"\\n\")\n",
    "textfile.close()\n",
    "merged_df\n",
    "\n",
    "# merged_df[merged_df[\"rep\"]==\"NODE_3_length_455224_cov_96.893901\"]\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "27f1f021",
   "metadata": {},
   "source": [
    "seqtk subseq assembled_contigs.fasta reference_genomes.txt > reference_genomes.fasta\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99407c21",
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly_df=pd.read_csv(\"/home/lmf/PhylloVir/SEQUENCING_TEST_VIROMES/VIRAL_FRACTION/08_ASSEMBLY_TEST/assembled_contigs.txt\",sep=\":\", names=[\"file\",\"contig\"])\n",
    "assembly_df[\"contig\"]=assembly_df[\"contig\"].str[1:]\n",
    "assembly_df[\"file\"]=assembly_df[\"file\"].str.split(\"_positive\").str[0]\n",
    "# merged_df.merge(assembly_df, right_on=\"contig\", left_on=)\n",
    "assembly_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4817d8be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "fasta_sequences = SeqIO.parse(open(\"/home/lmf/PhylloVir/SEQUENCING_TEST_VIROMES/VIRAL_FRACTION/08_ASSEMBLY_TEST/reference_genomes.fasta\"),'fasta')\n",
    "with open(\"/home/lmf/PhylloVir/SEQUENCING_TEST_VIROMES/VIRAL_FRACTION/reference_genomes_formatted.fasta\", \"w\") as out_file:\n",
    "    for fasta in fasta_sequences:\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "        print(\">\" + list(assembly_df[assembly_df[\"contig\"]==name][\"file\"])[0]+ \"_\"+ name, file=out_file)\n",
    "        print(sequence, file=out_file)\n",
    "        \n",
    "#         new_sequence = some_function(sequence)\n",
    "#         write_fasta(out_file)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f5f96ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat reference_full_phages.fasta reference_genomes_formatted.fasta > reference_phages.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79768a0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "# work_dir=\"/home/lmf/PhylloVir/VIRAL_WORLD/VIRAL_FRACTION\"\n",
    "# core_virome_df=pd.read_csv(work_dir + \"/core_metadata.csv\", index_col=0)\n",
    "# core_virome_df.rename({\"checkv_quality\": \"checkv_quality_votu\"}, axis=1, inplace=True)\n",
    "# # # color_plot_df\n",
    "# # checkv_df=pd.read_csv(\"/home/lmf/PhylloVir/VIRAL_WORLD/VIRAL_FRACTION/05_vOTUs/merged_checkV_quality_summary.txt\"\n",
    "# #                      ,sep=\"\\t\", names=[\"contig_id\", \"contig_length_votu\", \"provirus_votu\", \"proviral_length\", \"gene_count\", \"viral_genes\", \"host_genes\", \"checkv_quality_votu\", \"miuvig_quality\", \"completeness_votu\", \"completeness_method\", \"contamination\", \"kmer_freq\", \"warnings_votu\"])\n",
    "# # checkv_df=checkv_df[[\"contig_id\", \"contig_length_votu\", \"checkv_quality_votu\", \"completeness_votu\", \"provirus_votu\", \"warnings_votu\"]]\n",
    "# # core_virome_df=core_virome_df.merge(checkv_df, left_on=\"contig\", right_on=\"contig_id\", how=\"left\")\n",
    "# # core_virome_df = core_virome_df.drop('contig_id_x', \"contig, axis=1)\n",
    "\n",
    "\n",
    "# core_virome_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f3fca62",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "raw",
   "id": "ce1cacd4",
   "metadata": {},
   "source": [
    "#combine all reports for the assembled phages\n",
    "workdir=\"/home/lmf/RUTH\"\n",
    "workdir=\"/home/lmf/PhylloVir/SEQUENCING_TEST_VIROMES/VIRAL_FRACTION\"\n",
    "cd ${workdir}/05_vOTUs\n",
    "awk 'FNR>1' *checkV*/q* > merged_checkV_quality_summary.txt\n",
    "\n",
    "#combine all reports for the subassembly \n",
    "cd ${workdir}/08_ASSEMBLY_TEST\n",
    "awk 'FNR>1' *checkV*/q* > merged_checkV_quality_summary.txt\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a71a66ac",
   "metadata": {},
   "source": [
    "cd /home/lmf/PhylloVir/VIRAL_WORLD/VIRAL_FRACTION/08_ASSEMBLY_TEST\n",
    "\n",
    "#EXTRACT HIGH AND MEDIUM QUALITY GENOMES\n",
    "grep -h \"High\" *checkV*/q* | cut -f1 > high_quality_genomes.txt\n",
    "grep -h \"Medium\" *checkV*/q* | cut -f1 >> high_quality_genomes.txt\n",
    "\n",
    "#COMBINE ALL THE SUBASSEMBLIES AND EXTRACT M AND H QUAL\n",
    "cat *positive*fasta > assembled_contigs.fasta\n",
    "grep \">\" *positive*fasta > assembled_contigs.txt\n",
    "\n",
    "seqtk subseq assembled_contigs.fasta high_quality_genomes.txt > high_quality_genomes.fasta\n",
    "grep -c \">\" high_quality_genomes.fasta #21731\n",
    "grep -c \">\" ${workdir}/05_vOTUs/combined_positive_viral_contigs.tot.fasta #189539\n",
    "\n",
    "#COMBINE vOTUS WITH M AND H QUAL\n",
    "cat ${workdir}/05_vOTUs/combined_positive_viral_contigs.tot.fasta high_quality_genomes.fasta > high_quality_genomes_and_workflow.fasta\n",
    "grep -c \">\" high_quality_genomes_and_workflow.fasta #211270\n",
    "\n",
    "#CLUSTER AT SPECIES LEVEL\n",
    "makeblastdb -in high_quality_genomes_and_workflow.fasta -dbtype nucl -out high_quality_genomes_and_workflow.fasta\n",
    "blastn -query high_quality_genomes_and_workflow.fasta -db high_quality_genomes_and_workflow.fasta -outfmt '6 std qlen slen' -max_target_seqs 10000 -out high_quality_genomes_and_workflow.blastout -num_threads 144 #14:52-16:36 = 1h,44min\n",
    "\n",
    "python /home/lmf/Mosaic/mosaic/scripts/anicalc_checkv.py  -i high_quality_genomes_and_workflow.blastout -o high_quality_genomes_and_workflow.aniout\n",
    "python /home/lmf/Mosaic/mosaic/scripts/aniclust_checkv.py --fna high_quality_genomes_and_workflow.fasta --ani high_quality_genomes_and_workflow.aniout --out high_quality_genomes_and_workflow_rep.txt --min_ani 95 --min_tcov 85 --min_qcov 0    \n",
    "cut  -f1 high_quality_genomes_and_workflow_rep.txt > high_quality_genomes_and_workflow_rep_list.txt\n",
    "seqtk subseq high_quality_genomes_and_workflow.fasta high_quality_genomes_and_workflow_rep_list.txt > high_quality_genomes_and_workflow_rep.fasta\n",
    "grep -c \">\" high_quality_genomes_and_workflow_rep.fasta #111051\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db222a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqtk subseq high_quality_genomes_and_workflow.fasta subassembly_vOTU_clustering_rep_list.csv > subassembly_vOTU_clustering_rep.fasta\n",
    "cp subassembly_vOTU_clustering_rep.fasta 95-85_positive_viral_contigs.tot.fasta\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93fdfb0c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
