{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ee329eda",
   "metadata": {},
   "source": [
    "# FILTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d73961a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO   \n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn3\n",
    "import numpy as np \n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "673b4c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_df_counts_paired=snakemake.input.df_counts_paired\n",
    "input_merged_summary=snakemake.input.merged_summary\n",
    "input_vOTUs_prefiltered=snakemake.input.vOTUs_prefiltered\n",
    "input_vibrant_circular=snakemake.input.vibrant_circular\n",
    "input_vibrant_positive=snakemake.input.vibrant_positive\n",
    "input_vibrant_quality=snakemake.input.vibrant_quality\n",
    "input_vibrant_summary=snakemake.input.vibrant_summary\n",
    "input_virsorter_table=snakemake.input.virsorter_table\n",
    "input_virsorter_positive_list=snakemake.input.virsorter_positive_list\n",
    "input_genomad_virus_summary=snakemake.input.genomad_virus_summary\n",
    "input_genomad_plasmid_summary=snakemake.input.genomad_plasmid_summary\n",
    "input_genomad_viral_fasta=snakemake.input.genomad_viral_fasta\n",
    "input_genomad_viral_fasta_conservative=snakemake.input.genomad_viral_fasta_conservative\n",
    "#________\n",
    "output_summary=snakemake.output.summary\n",
    "output_filtered_list=snakemake.output.filtered_list\n",
    "\n",
    "SAMPLES=snakemake.params.samples\n",
    "min_votu_len=snakemake.params.min_votu_len\n",
    "SAMPLES_key=snakemake.params.key_samples\n",
    "input_mapping_dir=snakemake.params.mapping_dir\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f9130ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty list to hold individual DataFrames\n",
    "df_list = []\n",
    "\n",
    "# Loop over each sample and process data\n",
    "for sample in SAMPLES_key:\n",
    "    # Load unfiltered coverage stats\n",
    "    print(sample)\n",
    "    covstats_unf = pd.read_csv(f\"{input_mapping_dir}/STATS_FILES/bowtie2_{sample}_unfiltered_contigs.tot_covstats.txt\", sep=\"\\t\", index_col=0)\n",
    "    covstats_unf = covstats_unf[[f\"bowtie2_{sample}_unfiltered_contigs_tot_filtered Read Count\"]]\n",
    "    \n",
    "    covstats_unf.columns = [sample]\n",
    "    df_list.append(covstats_unf)\n",
    "\n",
    "# Concatenate all DataFrames at once along the columns axis\n",
    "combined_reads_u = pd.concat(df_list, axis=1)\n",
    "\n",
    "# Display or use the combined DataFrame\n",
    "combined_reads_u.head()\n",
    "\n",
    "# Calculate count DataFrame as before\n",
    "count_df = (combined_reads_u > 1).sum(axis=1).to_frame()\n",
    "mapped_contigs = list(count_df[count_df[0] > 0].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20598ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list to store vOTU names\n",
    "vOTU_list = []\n",
    "\n",
    "# Iterate over records in input_vOTUs_prefiltered file and append record IDs to the list\n",
    "for record in SeqIO.parse(input_vOTUs_prefiltered, \"fasta\"):\n",
    "    vOTU_list.append(record.id)\n",
    "\n",
    "# Create a DataFrame with vOTU names\n",
    "vOTU_prefiltered = pd.DataFrame(vOTU_list, columns=[\"vOTU_name\"])\n",
    "\n",
    "# Read data from input_merged_summary file into checkv_df DataFrame with specified column names\n",
    "checkv_df = pd.read_csv(input_merged_summary, sep=\"\\t\", names=[\"checkV_contig_id\", \"checkV_contig_length\", \"checkV_provirus\",\n",
    "                                                                \"checkV_proviral_length\", \"checkV_gene_count\", \"checkV_viral_genes\",\n",
    "                                                                \"checkV_host_genes\", \"checkv_quality\", \"checkV_miuvig_quality\",\n",
    "                                                                \"checkV_completeness\", \"checkV_completeness_method\", \"checkV_contamination\",\n",
    "                                                                \"checkV_kmer_freq\", \"checkV_warnings\"])\n",
    "checkv_df=checkv_df.groupby([\"checkV_contig_id\"]).first().reset_index()\n",
    "\n",
    "# Read data from input_vibrant_circular file into vibrant_circular_df DataFrame with a single column name\n",
    "vibrant_circular_df = pd.read_csv(input_vibrant_circular, names=[\"VIBRANT_circular_name\"])\n",
    "vibrant_circular_df[\"VIBRANT_circular\"] = \"Y\"\n",
    "\n",
    "# Read data from input_vibrant_positive file into vibrant_positive_df DataFrame with a single column name\n",
    "vibrant_positive_df = pd.read_csv(input_vibrant_positive, names=[\"VIBRANT_positive_name\"])\n",
    "vibrant_positive_df[\"VIBRANT_positive\"] = \"Y\"\n",
    "\n",
    "# Read data from input_vibrant_quality file into vibrant_quality_df DataFrame with specific column names\n",
    "vibrant_quality_df = pd.read_csv(input_vibrant_quality, sep=\"\\t\")\n",
    "vibrant_quality_df.columns = [\"VIBRANT_scaffold\", \"VIBRANT_lifecycle\", \"VIBRANT_quality\"]\n",
    "vibrant_quality_df = vibrant_quality_df.groupby(\"VIBRANT_scaffold\").first()\n",
    "\n",
    "# Read data from input_vibrant_summary file into vibrant_summary_df DataFrame with modified column names\n",
    "vibrant_summary_df = pd.read_csv(input_vibrant_summary, sep=\"\\t\")\n",
    "vibrant_summary_df.columns = ['VIBRANT_' + str(col).replace(\" \", \"_\") for col in vibrant_summary_df.columns]\n",
    "\n",
    "# Read data from input_virsorter_table file into virsorter_summary_df DataFrame with modified column names\n",
    "virsorter_summary_df = pd.read_csv(input_virsorter_table, sep=\"\\t\")\n",
    "virsorter_summary_df.columns = ['VirSorter2_' + str(col).replace(\" \", \"_\") for col in virsorter_summary_df.columns]\n",
    "virsorter_summary_df[\"VirSorter2_positive\"] = \"Y\"\n",
    "\n",
    "# Read data from input_genomad_virus_summary file into genomad_summary_df DataFrame with modified column names\n",
    "genomad_summary_df = pd.read_csv(input_genomad_virus_summary, sep=\"\\t\")\n",
    "genomad_summary_df.columns = ['geNomad_' + str(col).replace(\" \", \"_\") for col in genomad_summary_df.columns]\n",
    "\n",
    "# Create an empty list to store viral names\n",
    "viral_list = []\n",
    "\n",
    "# Iterate over records in input_genomad_viral_fasta file and append record IDs to the list\n",
    "for record in SeqIO.parse(input_genomad_viral_fasta, \"fasta\"):\n",
    "    viral_list.append(record.id)\n",
    "\n",
    "# Create a DataFrame with viral names\n",
    "geNomad_viral = pd.DataFrame(viral_list, columns=[\"geNomad_viral_name\"])\n",
    "geNomad_viral[\"geNomad_viral\"] = \"Y\"\n",
    "\n",
    "# Create an empty list to store conservative viral names\n",
    "viral_list_conservative = []\n",
    "\n",
    "# Iterate over records in input_genomad_viral_fasta_conservative file and append record IDs to the list\n",
    "for record in SeqIO.parse(input_genomad_viral_fasta_conservative, \"fasta\"):\n",
    "    viral_list_conservative.append(record.id)\n",
    "\n",
    "# Create a DataFrame with conservative viral names\n",
    "geNomad_viral_conservative = pd.DataFrame(viral_list_conservative, columns=[\"geNomad_viral_conservative_name\"])\n",
    "geNomad_viral_conservative[\"geNomad_viral_conservative\"] = \"Y\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0dbbc61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge vOTU_prefiltered with vibrant_circular_df based on vOTU_name and VIBRANT_circular_name columns\n",
    "df_temp1 = vOTU_prefiltered.merge(vibrant_circular_df, left_on=\"vOTU_name\", right_on=\"VIBRANT_circular_name\", how=\"left\")\n",
    "\n",
    "# Merge df_temp1 with vibrant_positive_df based on vOTU_name and VIBRANT_positive_name columns\n",
    "df_temp2 = df_temp1.merge(vibrant_positive_df, left_on=\"vOTU_name\", right_on=\"VIBRANT_positive_name\", how=\"left\")\n",
    "\n",
    "# Merge df_temp2 with vibrant_quality_df based on vOTU_name and VIBRANT_scaffold columns\n",
    "df_temp3 = df_temp2.merge(vibrant_quality_df, left_on=\"vOTU_name\", right_on=\"VIBRANT_scaffold\", how=\"left\")\n",
    "\n",
    "# Merge df_temp3 with vibrant_summary_df based on vOTU_name and VIBRANT_scaffold columns\n",
    "df_temp4 = df_temp3.merge(vibrant_summary_df, left_on=\"vOTU_name\", right_on=\"VIBRANT_scaffold\", how=\"left\")\n",
    "\n",
    "# Merge df_temp4 with virsorter_summary_df based on vOTU_name and VirSorter2_seqname columns\n",
    "df_temp5 = df_temp4.merge(virsorter_summary_df, left_on=\"vOTU_name\", right_on=\"VirSorter2_seqname\", how=\"left\")\n",
    "\n",
    "# Merge df_temp5 with genomad_summary_df based on vOTU_name and geNomad_seq_name columns\n",
    "df_temp6 = df_temp5.merge(genomad_summary_df, left_on=\"vOTU_name\", right_on=\"geNomad_seq_name\", how=\"left\")\n",
    "\n",
    "# Merge df_temp6 with geNomad_viral based on vOTU_name and geNomad_viral_name columns\n",
    "df_temp7 = df_temp6.merge(geNomad_viral, left_on=\"vOTU_name\", right_on=\"geNomad_viral_name\", how=\"left\")\n",
    "\n",
    "# Merge df_temp7 with geNomad_viral_conservative based on vOTU_name and geNomad_viral_conservative_name columns\n",
    "df_temp8 = df_temp7.merge(geNomad_viral_conservative, left_on=\"vOTU_name\", right_on=\"geNomad_viral_conservative_name\", how=\"left\")\n",
    "\n",
    "# Merge df_temp8 with checkv_df based on vOTU_name and checkV_contig_id columns\n",
    "df_merged = df_temp8.merge(checkv_df, left_on=\"vOTU_name\", right_on=\"checkV_contig_id\", how=\"left\")\n",
    "\n",
    "# Drop unnecessary columns from df_merged\n",
    "df_merged = df_merged.drop([\"VIBRANT_circular_name\", \"VIBRANT_positive_name\", \"VIBRANT_scaffold\", \"checkV_contig_id\"], axis=1)\n",
    "\n",
    "# Return the merged DataFrame\n",
    "df_merged\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69f50af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=df_merged[[\"VirSorter2_positive\",\"VIBRANT_positive\",\"geNomad_viral\",\"geNomad_viral_conservative\"]]\n",
    "\n",
    "# Get the counts for each category\n",
    "a_count = df['VirSorter2_positive'].value_counts()['Y']\n",
    "b_count = df['VIBRANT_positive'].value_counts()['Y']\n",
    "c_count = df['geNomad_viral'].value_counts()['Y']\n",
    "ab_count = df[df['VirSorter2_positive'] == 'Y']['VIBRANT_positive'].value_counts()['Y']\n",
    "ac_count = df[df['VirSorter2_positive'] == 'Y']['geNomad_viral'].value_counts()['Y']\n",
    "bc_count = df[(df['VIBRANT_positive'] == 'Y') & (df['geNomad_viral'] == 'Y')]['VIBRANT_positive'].value_counts()['Y']\n",
    "abc_count = df[(df['VirSorter2_positive'] == 'Y') & (df['VIBRANT_positive'] == 'Y') & (df['geNomad_viral'] == 'Y')]['VIBRANT_positive'].value_counts()['Y']\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn_labels = {'100': a_count - ab_count - ac_count + abc_count,\n",
    "               '010': b_count - ab_count - bc_count + abc_count,\n",
    "               '001': c_count - ac_count - bc_count + abc_count,\n",
    "               '110': ab_count - abc_count,\n",
    "               '101': ac_count - abc_count,\n",
    "               '011': bc_count - abc_count,\n",
    "               '111': abc_count}\n",
    "\n",
    "plt.figure(figsize=(8, 6))\n",
    "venn_diagram = venn3(subsets=venn_labels, set_labels=('VirSorter2', 'VIBRANT', 'geNomad'))\n",
    "\n",
    "# Add the counts to the Venn diagram\n",
    "for text in venn_diagram.set_labels:\n",
    "    text.set_fontsize(14)\n",
    "for text in venn_diagram.subset_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(12)\n",
    "\n",
    "# Show the Venn diagram\n",
    "# plt.title(\"Venn Diagram of Categories 'VirSorter2_positive', 'VIBRANT_positive', and 'geNomad_viral'\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b60fc8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define columns to count\n",
    "columns_to_count = [\"VirSorter2_positive\", \"VIBRANT_positive\", \"geNomad_viral\"]\n",
    "\n",
    "# Count the number of 'Y' occurrences in columns specified in columns_to_count and store the result in a new column 'count_Y'\n",
    "df_merged['count_Y'] = df_merged[columns_to_count].eq('Y').sum(axis=1)\n",
    "\n",
    "# Define the desired column order in the DataFrame\n",
    "move_cols = [\"vOTU_name\", \"count_Y\", \"checkV_contig_length\", \"VirSorter2_positive\", \"VIBRANT_positive\", \"geNomad_viral\", \"geNomad_viral_conservative\"]\n",
    "\n",
    "# Reorder the columns in df_merged according to move_cols and keep the remaining columns in their original order\n",
    "df_merged = df_merged[move_cols + [col for col in df_merged.columns if col not in move_cols]]\n",
    "\n",
    "# Create a filtered DataFrame by applying conditions on count_Y and checkV_contig_length columns\n",
    "df_merged_filtered =df_merged[\n",
    "    ((df_merged[\"count_Y\"] == 3) | (df_merged[\"geNomad_viral_conservative\"] == \"Y\")) & \n",
    "    ((df_merged[\"checkV_contig_length\"] >= min_votu_len) | (df_merged[\"checkV_miuvig_quality\"] == \"High-quality\")) ]\n",
    "\n",
    "df_merged_filtered_abundance=df_merged_filtered[(df_merged_filtered.vOTU_name.isin(mapped_contigs))]\n",
    "\n",
    "df_merged_removed=df_merged[\n",
    "    ((df_merged[\"count_Y\"] != 3) & (df_merged[\"geNomad_viral_conservative\"] != \"Y\")) |\n",
    "    ((df_merged[\"checkV_contig_length\"] < min_votu_len) & (df_merged[\"checkV_miuvig_quality\"] != \"High-quality\"))]\n",
    "\n",
    "df_merged_removed_abundance=df_merged_filtered[~(df_merged_filtered.vOTU_name.isin(mapped_contigs))]\n",
    "\n",
    "df_merged_removed=pd.concat([df_merged_removed,df_merged_removed_abundance])\n",
    "\n",
    "# Write the values in the 'vOTU_name' column to output_filtered_list file without the index and header\n",
    "df_merged_filtered_abundance[\"vOTU_name\"].to_csv(output_filtered_list, index=False, header=False)\n",
    "\n",
    "# Write df_merged_filtered_abundance to output_summary file\n",
    "df_merged_filtered_abundance.to_csv(output_summary)\n",
    "\n",
    "# df_merged_removed\n",
    "print(\"all\\t\\t\", len(df_merged))\n",
    "print(\"selected\\t\",len(df_merged_filtered_abundance))\n",
    "print(\"removed\\t\\t\",len(df_merged_removed))\n",
    "\n",
    "print(\"sum\\t\\t\",len(df_merged_filtered_abundance)+len(df_merged_removed) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "category_colors = [\"red\", \"blue\", \"green\", \"orange\", \"purple\"]\n",
    "\n",
    "df_merged_removed[\"checkV_contig_length_log\"]=np.log10(df_merged_removed[\"checkV_contig_length\"])\n",
    "\n",
    "\n",
    "# Plot the data as a distribution plot\n",
    "ax = sns.jointplot(data=df_merged_removed,x=\"checkV_contig_length_log\",y=\"checkV_completeness\", hue=\"checkv_quality\", alpha=1,height=6,palette=category_colors, marker=\"+\")\n",
    "ax.plot_joint(sns.kdeplot)\n",
    "\n",
    "# Set x- and y-axis labels  \n",
    "ax.ax_joint.set_xlabel('Contig length')\n",
    "ax.ax_joint.set_ylabel('CheckV completness')\n",
    "ax.ax_joint.set_ylim(0,100)\n",
    "ax.ax_joint.set_xlim(3,5)\n",
    "\n",
    "# Set the x-ticks and y-ticks to powers of 10\n",
    "# x-ticks\n",
    "xticks = ax.ax_joint.get_xticks()\n",
    "xticklabels = [int(10**x) for x in xticks]\n",
    "ax.ax_joint.set_xticklabels(xticklabels)\n",
    "\n",
    "# Set legend\n",
    "sns.move_legend(ax.ax_joint, title='Checkv quality',  markerscale=1, loc=\"lower right\",fontsize=10,title_fontsize=12)\n",
    "\n",
    "# # Save the figure as png and svg files\n",
    "# ax.figure.savefig(output_filtered_vOTUs_png, format=\"png\", bbox_inches = \"tight\",transparent=True)\n",
    "# ax.figure.savefig(output_filtered_vOTUs_svg, format=\"svg\", bbox_inches = \"tight\",transparent=True)\n",
    "# plt.setp(ax.get_legend().get_texts(), fontsize='22') # for legend text\n",
    "# plt.setp(ax.get_legend().get_title(), fontsize='32') # for legend title\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb97b76b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged_filtered_abundance.groupby(\"checkv_quality\").size().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40354aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "\n",
    "df_quality_counts=df_merged_filtered_abundance.groupby('checkv_quality').size()\n",
    "df_quality_counts=df_quality_counts.reindex(['Complete','High-quality', 'Medium-quality','Low-quality','Not-determined']).fillna(0)\n",
    "\n",
    "def make_autopct(values):\n",
    "    def my_autopct(pct):\n",
    "        total = sum(values)\n",
    "        val = int(round(pct*total/100.0))\n",
    "        return '{p:.2f}%  ({v:d})'.format(p=pct,v=val)\n",
    "    return my_autopct\n",
    "\n",
    "\n",
    "explode = (0.3, 0.2, 0.1, 0, 0)\n",
    "\n",
    "\n",
    "mycolors=[\"#d4dbba\",\"#ebeddd\",\"#372f69\",\"#efeff7\",\"#e0cf94\"]\n",
    "\n",
    "plt.pie(df_quality_counts, labels=df_quality_counts.index, autopct=make_autopct(df_quality_counts), explode=explode, textprops={'fontsize': 12}, colors=mycolors)\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
